File: Fortran90/GMRESm.f90
!----------------------------------------------------------------------
! Openpipeflow.org.  If used in your work, please cite
! Willis, A. (2017) SoftwareX 6, 124-127.
! https://doi.org/10.1016/j.softx.2017.05.003 (open access)
!                                      Thanks in advance! Ashley 2019.
!----------------------------------------------------------------------
! solve A x = b for x ;  
! minimise |Ax-b| subject to constraint |x| < delta .
! requires lapack routines dgelsy, dgesvd.
!----------------------------------------------------------------------
! m	  gmres dimension
! n 	  dimension of x
! x	  on input:  guess for x, can be 0
!         on exit:  solution x, subject to constraint if del>0
! b	  input b
! matvec  performs y := A x, call matvec(N,x, y)
! psolve  preconditioner, solve M x_out = x_in, call psolve(N,x)
! dotprd  dot product, d = dotprd(n,a,b)
! h       Hessian matrix,  size (m+1)*m
! v       Krylov subspace, size n*(m+1)
! res	  on input: |Ax-b|/|b|<res; 
!         on exit:  residual reached
! del     on input: if(del>0) then the x returned is the hookstep
!         on exit:  norm of next b predicted by hook
! its	  on input: max num its; 
!         on exit:  number of its taken
! info	  on input: if(info==1) print* residuals
!                   if(info==2) recalc hookstep with new del
! 	  on exit:  0 sucessful, 1 method breakdown, 2 max its
!							A.P.Willis 2008
!----------------------------------------------------------------------

 subroutine gmresm(m,n,x,b,matvec,psolve,dotprd,h,v,res,del,its,info)
   implicit none
   integer,          intent(in)    :: m
   integer,          intent(in)    :: n
   double precision, intent(inout) :: x(n)
   double precision, intent(in)    :: b(n)
   external                        :: matvec,psolve
   double precision, external      :: dotprd
   double precision, intent(inout) :: h(m+1,m)
   double precision, intent(inout) :: v(n,m+1)
   double precision, intent(inout) :: res
   double precision, intent(inout) :: del
   integer,          intent(inout) :: its
   integer,          intent(inout) :: info
   double precision :: tol,res_,stgn, w(n), z(n)
   double precision :: h_(m+1,m), y(m+1), p(m+1), work(4*m+1)
   integer :: imx, piv(m), rank, i
   double precision, save :: beta
   integer, save :: j
   logical :: done   

   if(info==2) then
      call hookstep(j,h,m,beta,del, y)
      z = matmul(v(:,1:j),y(1:j))
      call psolve(n, z)
      x = z
      info = 0
      return
   end if	 

   tol = res
   imx = its
   its = 0
   v   = 0d0

 1 continue
   res_ = 1d99
   stgn = 1d0 - 1d-14
 
   beta = dsqrt(dotprd(n,x,x)) 
   if(beta==0d0)  w = 0d0
   if(beta/=0d0)  call matvec(n,x, w)
   w = b - w
   beta = dsqrt(dotprd(n,w,w)) 
   v(:,1) = w / beta
     
   h = 0d0
   do j = 1, m
      its = its + 1
      z = v(:,j)      
      call psolve(n, z)
      call matvec(n,z, w)
      do i = 1, j
         h(i,j) = dotprd(n,w,v(1,i))
         w = w - h(i,j)*v(:,i)
      end do
      h(j+1,j) = dsqrt(dotprd(n,w,w))
      v(:,j+1) = w / h(j+1,j)
          
      p(1) = beta
      p(2:j+1) = 0d0
      h_(1:j+1,1:j) = h(1:j+1,1:j)
      call dgelsy(j+1,j,1,h_,m+1,p,m+1,piv,m,rank,work,4*m+1,i)
      if(i/=0) stop 'gmresm: dgelsy'
      y = p

      p(1:j+1) = - matmul(h(1:j+1,1:j),y(1:j))
      p(1) = p(1) + beta
      res = dsqrt(dot_product(p(1:j+1),p(1:j+1)))
      if(info==1) print*, 'gmresm: it=', its,' res=', real(res)
      
      done = (res<=tol .or. its==imx .or. res>res_)
      if(done .or. j==m) then
         if(del>0d0)  call hookstep(j,h,m,beta,del, y)
         z = matmul(v(:,1:j),y(1:j))
         call psolve(n, z)
         x = x + z
         if(its==imx) info = 2
         if(res>res_) info = 1
         if(res<=tol) info = 0
         if(done)     return
         if(del>0d0)  print*, 'gmres: warning! restart affects hookstep'
         goto 1       ! (j==m) restart
      end if
      res_ = res*stgn

   end do   
 
 end subroutine gmresm
 
 
!-----------------------------------------------------------------
! replace y with a vector that generates a hookstep
! c.f. Viswanath (2008) arXiv:0809.1498
!-----------------------------------------------------------------
 subroutine hookstep(j,h,m,beta,del, y)
   implicit none
   integer,          intent(in)    :: j, m
   double precision, intent(in)    :: h(m+1,j), beta
   double precision, intent(inout) :: del
   double precision, intent(out)   :: y(j)
   double precision :: a(j+1,j), s(j), u(j+1,j+1), vt(j,j), work(5*(j+1))
   double precision :: p(j+1), q(j), mu, qn
   integer :: info
   
   a = h(1:j+1,1:j)
   
   call dgesvd('A','A',j+1,j,a,j+1,s,u,j+1,vt,j,work,5*(j+1),info)
   if(info/=0) stop 'hookstep: dgesvd'
   
   p(1:j) = beta * u(1,1:j)   

   mu = max(s(j)*s(j)*1d-6,1d-99)
   qn = 1d99
   do while(qn>del)
      mu = mu * 1.1d0
      q = p(1:j)*s/(mu+s*s)
      qn = dsqrt(dot_product(q,q))
   end do

   y = matmul(q,vt)

   p = - matmul(h(1:j+1,1:j),y(1:j))
   p(1) = p(1) + beta
   del = dsqrt(dot_product(p,p))
 
 end subroutine hookstep

File: MATLAB/GMRESm.m
%----------------------------------------------------------------------
% Openpipeflow.org.  If used in your work, please cite
% Willis, A. (2017) SoftwareX 6, 124-127.
% https://doi.org/10.1016/j.softx.2017.05.003 (open access)
%                                      Thanks in advance! Ashley 2019.
%----------------------------------------------------------------------
% solve A x = b for x ;  
% minimise |Ax-b| subject to constraint |x| < delta .
% requires lapack routines dgelsy, dgesvd.
%----------------------------------------------------------------------
% m	  gmres dimension
% n 	  dimension of x
% x	  on input:  guess for x, can be 0
%         on exit:  solution x, subject to constraint if del>0
% b	  input b
% matvec  performs y = A x:  y = matvec(N,x)
% psolve  preconditioner, solve M y = x:  y = psolve(N,x)
% dotprd  dot product, d = dotprd(n,a,b)
% h       Hessian matrix,  size (m+1)*m
% v       Krylov subspace, size n*(m+1)
% res	  on input: |Ax-b|/|b|<res; 
%         on exit:  residual reached
% del     on input: if(del>0) then the x returned is the hookstep
%         on exit:  norm of next b predicted by hook
% its	  on input: max num its; 
%         on exit:  number of its taken
% info	  on input: if(info==1) print* residuals
%                   if(info==2) recalc hookstep with new del
% 	  on exit:  0 sucessful, 1 method breakdown, 2 max its
%----------------------------------------------------------------------

 function [x,res,del,its,info] = ...
                  GMRESm(m,n,x,b,matvec,psolve,dotprd,res,del,its,info)
   persistent h
   persistent v
   persistent beta_
   persistent j_

   if info == 2  
      [y,del] = GMREShook(j_,h,m,beta_,del) ;
      z = v(:,1:j_)*y(1:j_) ;
      x = psolve(n,z) ;
      info = 0 ;
      return
   end	 

   tol = res ;
   imx = its ;
   its = 0 ;
   v   = zeros(n,m+1) ;

   while 1 %(restart)
   res_ = 1d99 ;
   stgn = 1d0 - 1d-14 ;
 
   beta_ = sqrt(dotprd(n,x,x)) ;
   if beta_ == 0d0 
      w = 0d0 ;
   else
      w = matvec(n,x) ;
   end 
   w = b - w ;
   beta_ = sqrt(dotprd(n,w,w)) ;
   v(:,1) = w / beta_ ;
     
   h = zeros(m+1,m) ;
   for j = 1:m
      j_ = j ;
      its = its + 1 ;
      z = v(:,j) ; 
      z = psolve(n,z) ;
      w = matvec(n,z) ;
      for i = 1:j
         h(i,j) = dotprd(n,w,v(:,i)) ;
         w = w - h(i,j)*v(:,i) ;
      end
      h(j+1,j) = sqrt(dotprd(n,w,w)) ;
      v(:,j+1) = w / h(j+1,j) ;
          
      p = zeros(j+1,1) ;
      p(1) = beta_ ;
      h_(1:j+1,1:j) = h(1:j+1,1:j) ;
      y = pinv(h_)*p;
      
      p = - h(1:j+1,1:j)*y ;
      p(1) = p(1) + beta_ ;
      res  = sqrt(sum(p.*p)) ;
      if info==1 
          fprintf('gmresm: it=%i  res=%e\n', its, res) ; 
      end
      
      done = ((res <= tol) || (its == imx) || (res > res_)) ;
      if done || j==m
         if del > 0d0  
            [y,del] = GMREShook(j,h,m,beta_,del) ;
         end
         z = v(:,1:j)*y(1:j) ;
         z = psolve(n,z) ;
         x = x + z ;
         if its==imx
            info = 2 ;
         end
         if res>res_
            info = 1 ;
         end
         if res<=tol
            info = 0 ;
         end 
         if done 
            return 
         end
         if del>0d0   
            fprintf('gmres: WARNING: m too small. restart affects hookstep.\n');
         end
      end
      res_ = res*stgn ;

   end %for
   end %while 
 
 end
 

